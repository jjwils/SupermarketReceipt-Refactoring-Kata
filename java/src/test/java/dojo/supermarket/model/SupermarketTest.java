package dojo.supermarket.model;

import dojo.supermarket.ReceiptPrinter;
import org.approvaltests.combinations.CombinationApprovals;
import org.junit.jupiter.api.Test;

import static org.junit.jupiter.api.Assertions.assertEquals;

public class SupermarketTest {
    public static final Product TOOTHBRUSH = new Product("toothbrush", ProductUnit.Each);
    public static final Product APPLES = new Product("apples", ProductUnit.Kilo);

    // Todo: test all kinds of discounts are applied properly

    @Test
    public void tenPercentDiscount() {

        // ASSERT
//        assertEquals(4.975, receipt.getTotalPrice(), 0.01);
//        assertEquals(Collections.emptyList(), receipt.getDiscounts());
//        assertEquals(1, receipt.getItems().size());
//        ReceiptItem receiptItem = receipt.getItems().get(0);
//        assertEquals(apples, receiptItem.getProduct());
//        assertEquals(price, receiptItem.getPrice());
//        assertEquals(quantity * price, receiptItem.getTotalPrice());
//        assertEquals(quantity, receiptItem.getQuantity());
//
//        String printReceipt = new ReceiptPrinter().printReceipt(receipt);
        //Approvals.verify(getReceipt(APPLES,1.99, 2.5, SpecialOfferType.TenPercentDiscount, TOOTHBRUSH, 10.0));

        CombinationApprovals.verifyAllCombinations((product, price, quantity, specialOfferType, specialOfferProduct, specialOfferArgument, numberOfTimestoAdd) -> getReceipt(product, price, quantity, specialOfferType, specialOfferProduct, specialOfferArgument, numberOfTimestoAdd),
                new Product[]{APPLES},
                new Double[]{1.99},
                new Double[]{2.5, 3.0, 5.0, 1.0},
                new SpecialOfferType[]{SpecialOfferType.TenPercentDiscount, SpecialOfferType.ThreeForTwo, SpecialOfferType.TwoForAmount, SpecialOfferType.FiveForAmount},
                new Product[]{TOOTHBRUSH, APPLES},
                new Double[]{10.0},
                new Integer[]{2, 0});

    }

    private String getReceipt(Product product, double price, double quantity, SpecialOfferType specialOfferType, Product specialOfferProduct, double specialOfferArgument, int numberOfTimestoAdd) {
        SupermarketCatalog catalog = new FakeCatalog();
        catalog.addProduct(product, price);

        Teller teller = new Teller(catalog);
        teller.addSpecialOffer(specialOfferType, specialOfferProduct, specialOfferArgument);

        ShoppingCart cart = new ShoppingCart();
        for (int i = 0; i < numberOfTimestoAdd; i++) {
            cart.addItemQuantity(product, quantity);

        }


        // ACT
        String printReceipt = new ReceiptPrinter().printReceipt(teller.checksOutArticlesFrom(cart));
        return printReceipt;
    }


}
