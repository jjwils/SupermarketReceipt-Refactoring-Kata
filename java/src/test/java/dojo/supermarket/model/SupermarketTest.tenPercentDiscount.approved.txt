[Product{name='apples', unit=Kilo}, 1.99, 2.5, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500

Total:                              9.95
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500
10.0% off(apples)                  -1.00

Total:                              8.96
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500

Total:                              9.95
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500
3 for 2(apples)                    -1.99

Total:                              7.96
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500

Total:                              9.95
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500
2 for 10.0(apples)                 12.04

Total:                             21.99
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500

Total:                              9.95
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              4.98
  1.99 * 2.500
apples                              4.98
  1.99 * 2.500
5 for 10.0(apples)                  0.05

Total:                             10.00
 
[Product{name='apples', unit=Kilo}, 1.99, 2.5, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000

Total:                             11.94
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000
10.0% off(apples)                  -1.19

Total:                             10.75
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000

Total:                             11.94
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000
3 for 2(apples)                    -3.98

Total:                              7.96
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000

Total:                             11.94
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000
2 for 10.0(apples)                 18.06

Total:                             30.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000

Total:                             11.94
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              5.97
  1.99 * 3.000
apples                              5.97
  1.99 * 3.000
5 for 10.0(apples)                  0.05

Total:                             11.99
 
[Product{name='apples', unit=Kilo}, 1.99, 3.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000

Total:                             19.90
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000
10.0% off(apples)                  -1.99

Total:                             17.91
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000

Total:                             19.90
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000
3 for 2(apples)                    -5.97

Total:                             13.93
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000

Total:                             19.90
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000
2 for 10.0(apples)                 30.10

Total:                             50.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000

Total:                             19.90
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              9.95
  1.99 * 5.000
apples                              9.95
  1.99 * 5.000
5 for 10.0(apples)                  0.10

Total:                             20.00
 
[Product{name='apples', unit=Kilo}, 1.99, 5.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TenPercentDiscount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              1.99
apples                              1.99
10.0% off(apples)                  -0.40

Total:                              3.58
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TenPercentDiscount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, ThreeForTwo, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, ThreeForTwo, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TwoForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              1.99
apples                              1.99
2 for 10.0(apples)                  6.02

Total:                             10.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, TwoForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, FiveForAmount, Product{name='toothbrush', unit=Each}, 10.0, 0] => 
Total:                              0.00
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 2] => apples                              1.99
apples                              1.99

Total:                              3.98
 
[Product{name='apples', unit=Kilo}, 1.99, 1.0, FiveForAmount, Product{name='apples', unit=Kilo}, 10.0, 0] => 
Total:                              0.00
 
